os/kernel/binary_manager/binary_manager_load.c

```diff
@@ -138,93 +131,59 @@ static int binary_manager_load(int bin_idx)
 #ifdef CONFIG_OPTIMIZE_APP_RELOAD_TIME
 	binp = BIN_LOADINFO(bin_idx);
 	if (binp) {
-		valid_bin_count = 1;
-		latest_idx = 0;
-		snprintf(filepath[latest_idx], CONFIG_PATH_MAX, "%s/%s_%s", BINARY_DIR_PATH, BIN_NAME(bin_idx), BIN_VER(bin_idx));
+		bin_count = 1;
+		snprintf(filepath, CONFIG_PATH_MAX, "%s/%s_%s", BINARY_DIR_PATH, BIN_NAME(bin_idx), BIN_VER(bin_idx));
 	} else
 #endif
 	{
-		bin_name = BIN_NAME(bin_idx);
-		name_len = strlen(bin_name);
-
-		latest_ver = -1;
-		latest_idx = -1;
-		valid_bin_count = 0;
-
-		/* Open a directory for user binaries, BINARY_DIR_PATH */
-		dirp = opendir(BINARY_DIR_PATH);
-		if (!dirp) {
-			bmdbg("Failed to open a directory, %s\n", BINARY_DIR_PATH);
-			return ERROR;
-		}
-
-		file_idx = 0;
-
-		bmvdbg("Open a directory, %s\n", BINARY_DIR_PATH);
-
-		/* Read each directory entry */
-		for (;;) {
-			struct dirent *entryp = readdir(dirp);
-			if (!entryp) {
-				/* Finished with this directory */
-				break;
-			}
-			/* Read and Verify a binary file */
-			if (DIRENT_ISFILE(entryp->d_type) \
-				&& !strncmp(entryp->d_name, bin_name, name_len) && entryp->d_name[name_len] == '_') {
-				snprintf(filepath[file_idx], CONFIG_PATH_MAX, "%s/%s", BINARY_DIR_PATH, entryp->d_name);
-				ret = binary_manager_read_header(filepath[file_idx], &header_data[file_idx], true);
-				if (ret == OK) {
-					valid_bin_count++;
-					version = (int)atoi(header_data[file_idx].bin_ver);
-					bmvdbg("Found valid header in version %d\n", version);
-					if (version > latest_ver) {
-						latest_ver = version;
-						latest_idx = file_idx;
-						bmvdbg("Latest version %d, idx %d\n", latest_ver, latest_idx);
-					}
-				}
-				file_idx++;
-			}
-		}
-		closedir(dirp);
-
-		if (valid_bin_count == 0 || latest_idx < 0) {
-			bmdbg("No valid binary in fs : %s\n", BIN_NAME(bin_idx));
-			return ERROR;
-		}
+		bin_count = BIN_FILECNT(bin_idx);
 	}
+
 	do {
 #ifdef CONFIG_OPTIMIZE_APP_RELOAD_TIME
 		if (!binp)
 #endif
 		{
-			strncpy(load_attr.bin_name, header_data[latest_idx].bin_name, BIN_NAME_MAX);
-			load_attr.bin_size = header_data[latest_idx].bin_size;
-			load_attr.compression_type = header_data[latest_idx].compression_type;
-			load_attr.ram_size = header_data[latest_idx].bin_ramsize;
-			load_attr.stack_size = header_data[latest_idx].bin_stacksize;
-			load_attr.priority = header_data[latest_idx].bin_priority;
-			load_attr.offset = CHECKSUM_SIZE + header_data[latest_idx].header_size;
-			strncpy(load_attr.bin_ver, header_data[latest_idx].bin_ver, BIN_VER_MAX);
+			/* Read header data and Check crc */
+			snprintf(filepath, CONFIG_PATH_MAX, "%s/%s_%s", BINARY_DIR_PATH, BIN_NAME(bin_idx), BIN_FILEVER(bin_idx, BIN_USEIDX(bin_idx)));
+			ret = binary_manager_read_header(filepath, &header_data, true);
+			if (ret != OK) {
+				if (--bin_count > 0) {
+					bmdbg("Failed to read head %s, try to load another file\n", filepath);
```

Comment:
When failure, we will try to read header again, not try to load. So I think `Failed to read head %s, try to read another file` is better.